# Systematic analysis of transcriptional and post-transcriptional regulation of metabolism in yeast

[Gon√ßalves et al.](http://biorxiv.org/content/early/2016/06/23/057398)

Cells react to extracellular perturbations with complex and intertwined responses. Systematic identification of the regulatory mechanisms that control these responses is still a challenge and requires tailored analyses integrating different types of molecular data. 

Here we acquired time-resolved metabolomics measurements in yeast under salt and pheromone stimulation and developed a machine learning approach to explore regulatory associations between metabolism and signal transduction. Existing phosphoproteomics measurements under the same conditions and kinase-substrate regulatory interactions were used to estimate the enzymatic activity of signalling kinases. 

Our approach identified informative associations between kinases and metabolic enzymes capable of predicting metabolic changes. We extended our analysis to two studies containing transcriptomics, phosphoproteomics and metabolomics measurements across a comprehensive panel of kinases/phosphatases knockouts and time-resolved perturbations to the nitrogen metabolism, conveying a total of 143 unique conditions. Our approach accurately estimated the change in activity of transcription factors, kinases and phosphatases and these were capable of building predictive models to infer the metabolic adaptations of previously unseen conditions across different dynamic experiments. 

Time-resolved experiments were significantly more informative than genetic perturbations to infer metabolic adaptation. This difference may be due to the indirect nature of the associations and of general cellular states that can hinder the identification of causal relationships. This work provides a novel genome-scale integrative analysis to propose putative transcriptional and post-translational regulatory mechanisms of metabolic processes.
